### ISSUE TITLE
Should `--library --dynamic` change the setting of CHPL_LIB_PIC to `pic` if it is not already?

### ISSUE THREAD

**Issue by lydia-duncan** [chapel-lang/chapel#11401]
It seems as though `--library --dynamic` should at least emit a warning if `CHPL_LIB_PIC` is not set to `pic`, since shared libraries usually should also be built to be position independent.  It was suggested that we adjust chpl --dynamic to internally set CHPL_LIB_PIC=pic if it is not already set.  

This seems like it would allow the aforementioned warning to fall out naturally if there isn't a version of the runtime built with `CHPL_LIB_PIC`, but position independent code gives a performance penalty and it might be surprising to have this environment variable situationally changed out from under the user.

gcc's `-shared` flag does not imply `-fPIC` by default (but we don't have to make the same choices as gcc).

**Comment by mppf** [comment]
AFAIK GCC supports -fPIC separately from --shared because there are some low-level reasons one might want to create part of a shared library that is not position independent. However it is normal for code compiled for a shared library to be compiled with -fPIC, so it seems to me that this should be the default when `--shared` is requested.

**Comment by dlongnecke-cray** [comment]
Alternatively, a stopgap for this issue which would prevent user-visible linker errors about text relocations would simply be to have the compiler issue a fatal error when a dynamic library is compiled and `CHPL_LIB_PIC!=pic`.

**Comment by dlongnecke-cray** [comment]
Just wanted to also bring up that I think `--library-python` should almost certainly imply both `--dynamic` and `CHPL_LIB_PIC=pic`, at least for `linux64`.

**Comment by lydia-duncan** [comment]
`--library-python` does not require `--dynamic`, it only requires `CHPL_LIB_PIC=pic` on `linux64`.  This is in the documentation for `--library-python`: https://chapel-lang.org/docs/latest/technotes/libraries.html#prerequisites

**Comment by dlongnecke-cray** [comment]
The `CHPL_LIB_PIC` setting just controls whether or not the runtime and third-party libraries are compiled with `pic`, right?

It seems like "picness" is a property of the current `chpl` invocation (e.g. a Chapel code requires `pic` if the `--dynamic` flag was thrown).

Could we not just maintain a state for `CHPL_LIB_PIC` internally and keep track of whether or not the third-party and runtime have been built with it yet? If not, we can lazily rebuild them.

Then we no longer have to have the user set `CHPL_LIB_PIC`, we just grab what we need for the current session.

Maybe in the best-case scenario we can actually stop documenting `CHPL_LIB_PIC` entirely?

It feels weird to have the compiler force override something that the user may have set, but at the same time there are values of `CHPL_LIB_PIC` today which can lead to completely invalid compilations.

IMO we should let the `--dynamic` flag be the driver for whether or not something is position-independent or not.

If we are worried about hitting the user with a potential 5-10 minute session while we rebuild the runtime/third-party, maybe we can add a prompt or something (defaulting to yes) so they can cancel if they don't want to (though realistically, what else can they do?).

**Comment by bradcray** [comment]
> Could we not just maintain a state for CHPL_LIB_PIC internally and keep track of whether or not the third-party and runtime have been built with it yet? 

Correct me if I'm wrong, but I think we already do by embedding either `lib_pic-pic` or `lib_pic-none` into the `lib` and `build` directory trees?

> If not, we can lazily rebuild them.

I think that rather than lazily rebuilding or building-on-demand, this should result in our typical "no runtime for this configuration" style message where they'd then use John's new utility script to understand what it was missing and do the `make` that was lacking.  In part for consistency, in part to avoid a "my `chpl hello.chpl` compile took 5 minutes rather than 30 seconds!?!" type of reaction.

**Comment by lydia-duncan** [comment]
If I remember right, there are cases where it is okay to build dynamically without position independent code, so I would be inclined to warn rather than fail to compile, but it seems like the common case is to want position independent code with libraries and especially python libraries unless you do something really special when installing Python

**Comment by dlongnecke-cray** [comment]
Correct me if I'm off base, but I think we've encountered cases recently where users have stumbled into `CHPL_LIB_PIC` and their response was basically something like "why do I even have to do this at all? What is this?"

Brad is right that we maintain a path state for `pic` right now since it's part of `chplenv`. What I was (poorly) trying to ask was "can we still maintain that state even if we don't advertise it to the user?" And I'm guessing we probably can.

To me it seems right to associate `pic` code with the `--dynamic` flag, as otherwise I'm not really sure what that flag means (other than create a dynamic library? Which on most (all?) modern platforms requires totally relocatable code).

It would also open up the `--dynamic` flag to create `pic` executables as well, which seems intuitive to me.

---

Where I'm at right now in my thought-process is:

- Most (all?) shared libraries require PIC
- The user has to manually turn PIC on/off via an environment flag, which seems kind of cumbersome
- We might like the compiler to infer PIC when needed so that users don't have to bother
- It seems like the way to infer this is whether or not `--dynamic` is thrown, more or less (if you asked me to implement this that's probably what I'd do)
- If we just automatically/lazily rebuild runtime stuff as PIC when needed, we could incur a compilation time cost, which might be shocking to a user
- This leads to the argument that we should require the user to set `CHPL_LIB_PIC` before doing a rebuild, to confirm that they _want_ to do that
- However, it feels weird to have a user set a `chplenv` flag just once to rebuild the runtime only to have it subsequently completely ignored in favor of whether or not the compiler needs to generate PIC code (is there any precedent for this in the `chplenv` flags?)

---

It sounds like we don't want to hit users with a recompile-all-the runtimes silently scenario, and it also seems like people aren't in favor of just presenting a prompt.

Maybe we could establish a new semantics where `CHPL_LIB_PIC` only controls whether the runtime/third-party are built with PIC? When a user throws `--dynamic` and the compiler needs PIC it will emit a warning/error at that time telling the user they need to rebuild. I think we should probably rename the flag to something like `CHPL_RT_DEPS_PIC` so that it no longer implies that we're causing libraries to be generated with PIC.

This still has problematic semantics of: "this flag is on, but it doesn't actually control anything when on other than allow a new PIC version of our runtime/third-party to be built", but maybe documentation could alleviate that confusion? We could document the new semantics accordingly and emphasize that just because `CHPL_RT_DEPS_PIC=true` does not mean that the pic version of the runtime will be used (and vice versa, I suppose).

**Comment by mppf** [comment]
> Correct me if I'm off base, but I think we've encountered cases recently where users have stumbled into `CHPL_LIB_PIC` and their response was basically something like "why do I even have to do this at all? What is this?"

Yes, that was here -- https://chapel.discourse.group/t/when-is-setting-chpl-lib-pic-necessary-for-working-build/12339 -- which I "fixed" with PR #19785.

> Maybe we could establish a new semantics where `CHPL_LIB_PIC` only controls whether the runtime/third-party are built with PIC?

I think that here you are effectively saying "Let's not have the compiled code use PIC or not" and I'm having trouble understanding how that could work. Sure, we could make static vs dynamic drive pic-ness building a library. But the problem is that, AFAIK, we need the pic-ness of the generated code to match the pic-ness of the libraries linked to it. (Otherwise why are there failure modes like we saw with Arch? It might well work one way but not the other, though).

**Comment by dlongnecke-cray** [comment]
So what I'm suggesting is that `--dynamic` drives PIC for compiled code and the new flag drives only whether or not the runtime stuff has been built with pic or not.

So if I have `CHPL_RT_PIC=none` and build the runtime, I get the non-PIC version, and then if I try to throw `--dynamic` on my build, it will check the paths using `chplenv` and see that the PIC version of stuff hasn't been built using the path-for-platform magic. That's when it emits an error saying "please set `CHPL_RT_PIC=1` and rebuild the runtime. So the user does that, and now regardless of whether or not `--dynamic` is thrown they have the right runtime for their use.

It's kind of clunky because the value of the flag essentially only controls what version of the runtime/third-party you'll build, and not which one you're using. But it's the best compromise I could come up with, since others did not seem to be a fan of just prompting the user to rebuild everything as PIC.

Another option which I am personally very much a fan of is just compile both PIC and non-PIC versions of the runtime/third-party. Apparently that cost might be prohibitive, but I think only developers will be recompiling often.
